### ISSUE TITLE
Make it an option to have a different mask for each sequence in a batch for the Transformer

### ISSUE THREAD

**Issue by erikhuck** [pytorch/pytorch#30678]
## 🚀 Feature

Please make it so we can have different masks for each sequence in a batch as input into a multi-head attention layer. Let's say you have a batch of sequences of shape [S, N, D]. I want to be able to use N different masks.

## Motivation

I want to be able to have different masks for each sequence in a batch of sequences. I'm referring to the input to a transformer and ultimately the input into a multi-head attention layer.

## Pitch

I've made my own masks for the attention mechanism of a transformer. That's the attention mask or "attn_mask" as its called in the multi_head_attention_forward function in torch.nn.functional. In the documentation, the required shape of the attention mask does not include the batch dimension and that's why it needs to unsqueeze(0) on line 3351. But I do not want to broadcast because not all the masks in the batch are the same. For my purposes, removing that line would work but to generalize you could check to see if attn_mask already has a batch dimenstion (i.e len(attn_mask.shape) > 2). And if it does, do not unsqueeze it but leave it alone.

## Alternatives

The only thing I can do is change the source code. As long as that line of code that unsqueezes is there, I can't do what I need to do. If that line was optional, I could.

## Additional context

 Here are the lines of code in torch.nn.functional.py, where 3351 is the problem.
3350     if attn_mask is not None:
3351     indent>attn_mask = attn_mask.unsqueeze(0)
3352     indent>attn_output_weights += attn_mask

An alternative, only requiring one extra line of code, that would work for me and others who want to have different masks for each sequence could be this:
3350     if attn_mask is not None:
3351     indent>if len(attn_mask.shape) < 3:
3352     indent>indent>attn_mask = attn_mask.unsqueeze(0)
3353     indent>attn_output_weights += attn_mask

**Comment by zou3519** [comment]
cc @zhangguanheng66 what do you think about this?

**Comment by zhangguanheng66** [comment]
Thanks for the proposal. Yes, it's all the way to `nn.MultiheadAttention`. We have a plan to refactor nn.MultiheadAttention by splitting the long function into several pieces. And we can make an option to accept a 3-D mask there.

**Comment by zhangguanheng66** [comment]
@erikhuck Please take a look at the fix https://github.com/pytorch/pytorch/pull/31996

**Comment by erikhuck** [comment]
@zhangguanheng66 That looks great to me. Thank you very much.
